selftest/reuse-fd/main.go

```diff
@@ -0,0 +1,91 @@
+package main
+
+import "C"
+
+import (
+	"bytes"
+	"log"
+	"unsafe"
+
+	bpf "github.com/aquasecurity/libbpfgo"
+)
+
+func main() {
+	bpfModule, err := bpf.NewModuleFromFile("main.bpf.o")
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer bpfModule.Close()
+
+	bpfModule.BPFLoadObject()
+
+	testerMap, err := bpfModule.GetMap("tester")
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	testerReusedMap, err := bpfModule.GetMap("tester_reused")
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	//
+	// BPFMap ReuseFD
+	//
+
+	// The current instance of "tester_reused" will be closed, and testerReusedMap
+	// will now point to the same map as testerMap "tester", however, via a
+	// different FD.
+	err = testerReusedMap.ReuseFD(testerMap.FileDescriptor())
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	key1 := uint32(1)
+	value1 := struct{ x int }{50}
+	key1Unsafe := unsafe.Pointer(&key1)
+	value1Unsafe := unsafe.Pointer(&value1)
+	err = testerMap.Update(key1Unsafe, value1Unsafe) // update "tester"
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	key2 := int64(42069420)
+	value2 := []byte{'a', 'b', 'c'}
+	key2Unsafe := unsafe.Pointer(&key2)
+	value2Unsafe := unsafe.Pointer(&value2[0])
+	err = testerReusedMap.Update(key2Unsafe, value2Unsafe) // also update "tester"
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	//
+	// BPFMapLow ReuseFD
+	//
+
+	toReuseCreated, err := bpf.CreateMap(bpf.MapTypeArray, "toreuse", 4, 4, 420, nil)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// The current instance of "toreuse" will be closed, and toReuseCreated
+	// will now point to the same map as testerMap "tester", however, via a
+	// different FD.
+	err = toReuseCreated.ReuseFD(testerMap.FileDescriptor())
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	val2, err := toReuseCreated.GetValue(key2Unsafe) // lookup "tester"
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if bytes.Equal(val2, value2) {
+		log.Fatal("wrong value")
```

Comment:
Shouldn't val2 and value2 be equal (since both are taken from tester map?)

Reply:
Yes. Good catch. It is a leftover from a previous test that passed through. I'm fixing this test. Thanks.